### PR TITLE
Feature refactor apply v2

### PR DESCRIPTION
**Pull request by sopel39**
Supercedes: https://github.com/prestodb/presto/pull/6701

FYI: @martint

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 1]
```diff
@@ -37,25 +41,43 @@
      */
     private final List<Symbol> correlation;
 
+    /**
+     * Assignments that operate on subqueries (e.g: "expr IN (SELECT 1)").
+     * It might be a direct mapping to a subquery symbol if subquery is scalar.
+     */
+    private final Map<Symbol, Expression> subqueryAssignments;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: What are the semantics of these assignments? What is the type of the symbol?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 2]
```diff
@@ -37,25 +41,43 @@
...
         this.input = input;
         this.subquery = subquery;
+        this.subqueryAssignments = subqueryAssignments;
         this.correlation = ImmutableList.copyOf(correlation);
     }
 
+    /**
+     * @return true when subquery is scalar and it's output symbols are directly mapped to ApplyNode output symbols
+     */
+    public boolean isResolvedScalarSubquery()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I would move that method to `ScalarQueryUtil`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -5611,6 +5611,9 @@ public void testSemiJoin()
                 "SELECT * FROM (VALUES (1,1), (2,2), (3, 3)) t(x, y) WHERE (x+y in (VALUES 4, 5)) AND (x*y in (VALUES 4, 5))",
                 "VALUES (2,2)");
 
+        // test multi level IN subqueries
+        assertQuery("SELECT 1 IN (SELECT 1), 2 IN (SELECT 1) WHERE 1 IN (SELECT 1)");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Can you add the same test to `TestLogicalPlanner` to check if there is only two semi joins?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 4]
```diff
@@ -197,29 +197,16 @@ private PlanBuilder appendInPredicateApplyNode(PlanBuilder subPlan, InPredicate
...
         checkState(inPredicate.getValueList() instanceof SubqueryExpression);
         PlanNode subquery = createRelationPlan(((SubqueryExpression) inPredicate.getValueList()).getQuery()).getRoot();
-        Map<Expression, Symbol> correlation = extractCorrelation(subPlan, subquery);
-        if (!correlationAllowed && correlation.isEmpty()) {
-            throwNotSupportedException(inPredicate, "Correlated subquery in given context");
-        }
-        subPlan = subPlan.appendProjections(correlation.keySet(), symbolAllocator, idAllocator);
-        subquery = replaceExpressionsWithSymbols(subquery, correlation);
 
-        TranslationMap translationMap = subPlan.copyTranslations();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Changes in this class does not match the commit message, probably they need to be done in separate commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RemoveUnreferencedScalarInputApplyNodes.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<PlanNode>
+    {
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            if (node.getInput().getOutputSymbols().isEmpty() && isScalar(node.getInput()) && node.isResolvedScalarSubquery()) {
+                return context.rewrite(node.getSubquery());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RemoveUnreferencedScalarInputApplyNodes.java]: should not be `node.getInput()` here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.GREATER_THAN;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class TransformQuantifiedComparisonApplyToScalarApply
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: this belongs to different commit and it is not used in this one

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: name is wrong it handles the EXISTS not quantified comparisons

**Comment by kokosing** [comment]
@martint I noticed that issue (witn IN) also and I didn't had any good idea how to fix that for a long time. Until @sopel39 come out with a concept of putting an Expression into the ApplyNode. 

Now semantic of ApplyNode changes, it no longer works as cross join (each input row merged with each subquery row). We decided it should now return single row for every input row and for each such row it should evaluate a subquery expression (x IN (SELECT ..) into a single symbol (value). 

Pros:
 - that way we exactly know which ApplyNode handles which subquery. 
 - when subquery symbol is pruned whole subquery experession and possibly apply node can  be simply pruned. We already struggled with this couple of times.
 - different subqueries can be handled in different optimizers (not like today were quantified comparisons and EXISTS have to be handled completely in `SubqueryPlanner`)

**Comment by martint** [comment]
> Now semantic of ApplyNode changes, it no longer works as cross join (each input row merged with each subquery row). We decided it should now return single row for every input row and for each such row it should evaluate a subquery expression (x IN (SELECT ..) into a single symbol (value).

That makes sense. That's how I originally though apply was supposed to work, which is similar to how a "map" function over an array would work in common functional languages.

So, if I'm understanding this correctly, a subquery expression assignment `a = x` is only valid if the subquery is guaranteed to produce a single row, no? If the subquery produces no rows, does apply work like an outer join?

BTW, we should add that explanation to the commit message. It was not at all clear that this was a major semantic change for that node type.

**Comment by kokosing** [comment]
Typically, subquery expression should be just `a` (vs `x = a`) and it is a scalar subquery, node which produces `a` should be wrapped in `EnforceSingleRow`, so there should be a value. `x = a` should be evaluated in `FilterNode` (or `ProjectNode`) afterwards.

In this case, when is no value (subquery plan was not wrapped with `EnforceSingleRow`) I would better raise an exception.

Anyway, I think it is a responsibility of `ApplyNode` how `a` is calculated in some expressions (like `EXISTS`) it may accept that subquery produces no value. But in case of `x = a` I think it is a bug.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: `ScalarQueryUtil` returns true if plan node returns single node. `ApplyNode` can have scalar subquery, but not be scalar, so it doesn't quite fit there.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 4 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Those are atomic changes unfortunately. Semantics of `ApplyNode` does not fit with the current `InPredicate` support. `ApplyNode` cannot output symbols that represent set (like it was before for in predicates). Therefore those changes should go together.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RemoveUnreferencedScalarInputApplyNodes.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RemoveUnreferencedScalarInputApplyNodes.java]: nope. input node is scalar with not symbols, therefore we can skip it altogether and just leave subquery node

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java (hunk 6 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToScalarApply.java]: moved to correct commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: Definitely I will extend this javadoc

**Comment by sopel39** [comment]
Let me clarify the concept a bit. The background is that in plan you have expressions like:
`x IN (subquery)`
`x = (subquery)`
`x = ALL (subquery)`
etc...
Those are expressions that produce a single value and you can use them in different context (e.g: in projection and filter) in SQL. However they are different than non-subquery expressions in such a way that their input is set of rows rather than a single row. Therefore subquery expressions cannot be placed in arbitrary place in logical plan (e.g: projection or filter). In logical plan projection or filter (linear operators) you cannot have expression `x IN (subquery_symbol)` since `subquery_symbol` is not a reference to an atomic value but rather a set of values. Such expression is invalid in logical plan. 

This is why I have introduced `subqueryAssignments` to `ApplyNode`. Subquery expressions are logically invalid in linear operators, but they fit perfectly in `ApplyNode`. The idea is that in `ApplyNode` context subquery expression can operate on symbols outputted by subquery (e.g: `x IN (subquery_symbol`), but they still produce single atomic value for each input row.

Examples:
```
ApplyNode
  - expression: in_symbol -> X IN (Y)
  - input output symbols: X
  - subquery output symbols: Y
  - ApplyNode output symbols: in_symbol, X, Y
  - meaning: in_symbol -> if set of Y values contains X
```

```
ApplyNode
  - expression: all_symbol -> X = ALL (Y)
  - input output symbols: X
  - subquery output symbols: Y
  - ApplyNode output symbols: all_symbol, X, Y
  - meaning: all_symbol -> if set of y values contains X
```
```
ApplyNode 
 - expression: Y -> Y
 - input output symbols: X
 - (SCALAR) subquery output symbols: Y
 - ApplyNode output symbols: Y, X
 - meaning: subquery is scalar (produces single row) therefore symbol Y can be used linearly (as any other symbol)
```

**Comment by sopel39** [comment]
applied comments, added javadoc (fixup commit), back to yoy @kokosing

**Comment by sopel39** [comment]
rebased on newest master

**Comment by martint** [comment]
Merged, thanks!
